pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1309,6 +1351,17 @@ func (og *operationGenerator) GenerateUnmapDeviceFunc(
 			return volumetypes.NewOperationContext(eventErr, detailedErr, migrated)
 		}
 
+		// Mark the device as uncertain to make sure kubelet calls UnmapDevice again in all the "return err"
+		// cases below. The volume is marked as fully un-mapped at the end of this function, when everything
+		// succeeds.
+		markDeviceUncertainErr := actualStateOfWorld.MarkDeviceAsUncertain(
+			deviceToDetach.VolumeName, deviceToDetach.DevicePath, globalMapPath)
+		if markDeviceUncertainErr != nil {
+			// On failure, return error. Caller will log and retry.
+			eventErr, detailedErr := deviceToDetach.GenerateError("UnmapDevice.MarkDeviceAsUncertain failed", markDeviceUncertainErr)
+			return volumetypes.NewOperationContext(eventErr, detailedErr, migrated)
+		}
+
```

Comment:
I see that we are marking device as uncertain even before unstage operation has started. This means that if `TearDownDevice` fails for a reason that is final (i.e which tells us unstage never really started) - then volume will still be treated as uncertain and hence a full stage has to be performed to make the volume usable again. 

This seems very different from how we handled uncertain devices/volumes in other places.

Reply:
another way is perform teardown first. In case of error returned, mark it as uncertain. In case of no error, mark it as unmounted.

Reply:
I pointed out in the description of the PR, IMO it's not safe to assume that a mount is usable even after UnmountVolume / UnmountDevice final error. Better to call SetUp / MountDevice again when the mount is needed. It is a corner case already and one extra MountDevice / SetUp won't harm, especially if it's idempotent.

I am marking the volume as uncertain at the beginning, because unmapping the volume is a long process - see below - and it makes the error handling much easier.

Reply:
I think my main concern is - other intree volume types do not benefit from being added to ASOW in uncertain state. Ideally all of `MountDevice` calls should be idempotent and hopefully nothing would break but it is somewhat of risky behaviour change and I am not sure if we have e2es that verify stuff like this. 

May be it is fine and may be - we will see some regressions.

Reply:
If any of the operation below fails (like `isDeviceOpened` check), the volume is still marked as mounted, while e.g. iscsi or FC volumes were already detached in `TearDownDevice`.  Therefore kubelet should mark it as uncertain. It is not related to CSI.